src/mbr.c

```diff
@@ -0,0 +1,329 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <glib/gstdio.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/mount.h>
+
+#include "mbr.h"
+#include "update_handler.h"
+
+/* partition entry in MBR partition table, the system boots from */
+#define BOOT_PARTITION_ENTRY		0
+#define MBR_SIZE			512
+#define MBR_FIRST_PART_ADDR		0x1BE
+#define MBR_NUMBER_OF_PARTITIONS	4
+#define MBR_MAGIC_NUMBER_ADDR		0x1FE
+#define MBR_MAGIC_NUMBER_L		0x55
+#define MBR_MAGIC_NUMBER_H		0xAA
+
+#define DEFAULT_SECTOR_SIZE		512
+#define HDIO_GETGEO			0x0301
+
+struct hd_geometry {
+	unsigned char heads;
+	unsigned char sectors;
+	unsigned short cylinders;
+	unsigned long start;
+};
+
+struct partition_tbl_entry {
+	guint8 boot_indicator;
+	guint8 chs_begin[3];
+	guint8 type;
+	guint8 chs_end[3];
+	guint32 partition_start;
+	guint32 partition_size;
+};
+
+static gboolean get_sectorsize(const gchar *device, guint *sector_size)
+{
+	gboolean res = FALSE;
+	guint arg;
+	gint fd;
+
+	fd = open(device, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	if (ioctl(fd, BLKSSZGET, &arg) != 0)
+		goto out;
+
+	*sector_size = arg;
+	res = TRUE;
+out:
+	if(fd >= 0)
+		close(fd);
+
+	return res;
+}
+
+static gboolean get_heads_and_sectors(const gchar *device, guchar *heads,
+		guchar *sectors, GError **error)
+{
+	gboolean res = FALSE;
+	gint fd;
+	struct hd_geometry geometry;
+
+	fd = open(device, O_RDONLY);
+	if (fd < 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"open %s failed: %s",
+				device, g_strerror(errno));
+		goto out;
+	}
+
+	if (ioctl(fd, HDIO_GETGEO, &geometry) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"ioctl command 0x%04x failed: %s",
+				HDIO_GETGEO, g_strerror(errno));
+		goto out;
+	}
+
+	*heads = geometry.heads;
+	*sectors = geometry.sectors;
+	res = TRUE;
+out:
+	if (fd >= 0)
+		close(fd);
+
+	return res;
+}
+
+static gboolean r_mbr_read_partitions(const gchar *device,
+		struct partition_tbl_entry *part, GError **error)
+{
+	gboolean res = FALSE;
+	FILE *f = NULL;
+
+	f = g_fopen(device, "r");
+
+	fseek(f, MBR_FIRST_PART_ADDR, SEEK_SET);
+
+	if (fread(part, sizeof(*part), MBR_NUMBER_OF_PARTITIONS, f) !=
+	    MBR_NUMBER_OF_PARTITIONS) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"read failed: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
+	res = TRUE;
+out:
+	if (f)
+		fclose(f);
+
+	return res;
+}
+
+static gboolean mbr_is_part_overlapping(guint32 start1, guint32 size1, guint32 start2,
+		guint32 size2)
+{
+	guint32 end1 = start1 + size1 - 1;
+	guint32 end2 = start2 + size2 - 1;
+
+	if(start1 >= start2 && start1 <= end2)
+		return TRUE;
+
+	if(start2 >= start1 && start2 <= end1)
+		return TRUE;
+
+	return FALSE;
+}
+
+gboolean r_mbr_get_non_active_boot_partition(const gchar *device,
+		struct mbr_partition *partition,
+		GError **error)
```

Comment:
s/non_active/inactive

I am not sure if it wouldn't be better to just return a slot id here instead of the actual addresses. This would separate the addresses from the more abstract concept of slots and simplify the update of the boot slot. However, as the caller needs the addresses anyway for writing the partition, we would need another function to get the addresses of a slot. Therefore, returning the partition address is probably fine here and makes things simpler.

Reply:
Yes, returning addresses here is fine. It's still prefer inactive over non_active.